File: ConvertSortedArrayToBinarySearchTree.java
/**
 * Author: Chih-Jye Wang
 * Date  : Dec 17, 2013
 *
 * Given an array where elements are sorted in ascending order, convert it to a height balanced BST.
 */
public class ConvertSortedArrayToBinarySearchTree
{
    public TreeNode sortedArrayToBST(int[] num)
    {
        return sortedArrayToBST(num, 0, num.length - 1);
    }
    
    TreeNode sortedArrayToBST(int[] a, int start, int end)
    {
        if(end < start || start > end)
            return null;
        if(start == end)
            return new TreeNode(a[start]);
        int mid = (start + end) / 2;
        TreeNode node = new TreeNode(a[mid]);
        node.left  = sortedArrayToBST(a, start, mid - 1);
        node.right = sortedArrayToBST(a, mid + 1, end);
        return node;
    }
}

File: ruby/convert_sorted_array_to_binary_search_tree.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {Integer[]} nums
# @return {TreeNode}
def sorted_array_to_bst(nums)
    return if !nums || nums.length == 0
    return bst(nums, 0, nums.length - 1)
end

def bst(nums, i, j)
    return if i > j
    mid = (j - i) / 2 + i
    node = TreeNode.new(nums[mid])
    node.left = bst(nums, i, mid - 1)
    node.right = bst(nums, mid + 1, j)
    return node
end